### DISCUSSION TITLE
Support more than one shared page

### DISCUSSION THREAD

**Discussion by dr-sandipan-chowdhury** [zauberzeug/nicegui#3265] [Ideas / Feature Requests]
### Question

How to make a private page annotated with ui.page as public? Is this possible. The usecase for this is: broadcasting video or music at a certain point of time for a certain duration. nicegui really makes this easy for unannotated public page but I want multiple of these public pages, how to acheive that?

Also, I get that an approach using binding actions to app.storage.general is avaiable but that is a little complex than public page

**Comment by falkoschindler** [comment]
Hi @bin-san,

It is currently not officially supported to have more than one shared page. But in principle it should be possible to create a shared client like this (which somehow doesn't work...):
```py
with Client(ui.page('/shared'), shared=True):
    ui.label(f'Shared ID: {uuid.uuid4()}')
```

I'll convert this discussion into a feature request. So maybe we can add support for multiple shared pages in the future.

**Comment by me21** [comment]
You can make the page act like shared by storing the relevant page state in a shared variable.

**Reply by petergaultney** [reply]
can you show a snippet of code? i'm not sure how to connect a page to a shared variable.

**Reply by me21** [reply]
Something like this:

```python
from nicegui import ui

class Demo:
    def __init__(self):
        self.number = 1

demo = Demo()

@ui.page('/')
def main_page():
    v = ui.checkbox('visible', value=True)
    with ui.column().bind_visibility_from(v, 'value'):
        ui.slider(min=1, max=3).bind_value(demo, 'number')
        ui.toggle({1: 'A', 2: 'B', 3: 'C'}).bind_value(demo, 'number')
        ui.number().bind_value(demo, 'number')

ui.run()
```

I haven't run this code but I expect that the number will be shared across all instances of the page, because it's bound to the field of the global variable. Checkbox state and widgets visibility, on another hand, can be controlled individually for each instance.

**Reply by petergaultney** [reply]
My concern with this approach would be that I don't know whether interpreter state is shared across all client connections. I assume starlette has some kind of underlying concurrency/parallelism strategy, and if it involves multiprocessing, I would tend to think that this Demo object would be re-created in every separate process upon module import, and therefore would not be shared by all connections.

**Reply by me21** [reply]
I'm not sure there's multiprocessing enabled by default. Maybe @falkoschindler can shed some light on whether it's a good approach or not.

**Reply by petergaultney** [reply]
from the research I've been doing, your approach would probably work by default - it's `uvicorn` that handles spawning multiple workers, and I believe by default it uses only 1. 

However, if someone passed `ui.run(...., workers=N`) (N>1) then this shared-memory approach would not work across all visitors to the page, as they might end up in separate processes.

I think in _theory_ it might be possible to solve this with `app.storage.general` and `bind_value` - I may have to test it to see whether it actually works across multiple workers, though.

**Reply by me21** [reply]
According to this: https://github.com/zauberzeug/nicegui/discussions/1539, multiprocessing is not fully supported by NiceGUI, if I understand correctly.

@falkoschindler, @rodja, what do you think about this approach to multiple shared pages?

**Reply by rodja** [reply]
Parallel execution should not be a problem. We use asyncio which does not need multiple processes to handle concurrency. Using multiple workers is a bad idea. They would not share the same client ids and depending on the load balancer you will get connection errors.
Multiple shared pages are possible. Here is a brief demo:

```py
from fastapi import Request
form nicegui import ui, app

ui.label('Shared Main Page')
ui.button('Click me', on_click=lambda: ui.label('Main!'))

with Client(ui.page('/hello'), request=None) as shared_hello:
    ui.button('Click me', on_click=lambda: ui.label('Hello World!'))

@app.get('/hello')
def hello_page(request: Request) -> Response:
    return shared_hello.build_response(request)
```

**Reply by falkoschindler** [reply]
> I think in theory it might be possible to solve this with `app.storage.general` and `bind_value`

I don't think this will work, because the storage module doesn't observe file changes, but only reads them once during startup. So it won't be useful for synchronizing state across processes.

**Comment by petergaultney** [comment]
If I'm understanding the discussion correctly, something like this would be extremely useful to me. 

I am building a method of coordinating between users, but if I could simply share the page state (or, ideally, some _subsection_ of the elements) between all visitors to a given route, then i would have to do a lot less work 😅 . Even if that mechanism were subject to lag - and even though I take it for granted that, as a distributed systems problem, there's no way to fundamentally _coordinate_ (i.e. merge) edits to the page state amongst multiple users without using something like CRDTs or OT (both of which are very much out of scope for a web framework like nicegui)
